scripts/polyrith_sage.py

```diff
@@ -0,0 +1,136 @@
+        return const_string(0)
+    out = []
+    for i in range(len(coeffs)):
+        if str(monomials[i]) == "1":
+            n = QQ(float(coeffs[i]))
+            out.append(const_string(n))
+        elif str(coeffs[i]) == "1":
+            out.append(monomial_to_string(monomials[i]))
+        elif str(coeffs[i]) == "-1":
+            out.append("-" + monomial_to_string(monomials[i]))
+        else:
+            out.append("(poly.mul " + const_string(coeffs[i]) + " " + monomial_to_string(monomials[i]) + ")")
+    return poly_terms_string(out)
+'''
+
+# future extensions may change behavior depending on the base type
+def type_str(type):
+    return "QQ"
+
+def var_names(var_list_string):
+    return var_list_string[1:-1].replace(" ", "")
+
+def create_query(type: str, var_list, eq_list, goal_type):
+    """ Create a query to invoke Sage's `MPolynomial_libsingular.lift`. See
+    https://github.com/sagemath/sage/blob/f8df80820dc7321dc9b18c9644c3b8315999670b/src/sage/rings/polynomial/multi_polynomial_libsingular.pyx#L4472-L4518
+    for a description of this method. """
+    query = f'''
+{var_names(var_list)} = {type_str(type)}['{var_names(var_list)}'].gens()
+gens = {eq_list}
+p = {goal_type}
+I = ideal(gens)
+coeffs = p.lift(I)
+print(list(map(polynomial_to_string, coeffs)))
+'''
+    return query
+
+class EvaluationError(Exception):
+    def __init__(self, ename, evalue, message='Error in Sage communication'):
+        self.ename = ename
+        self.evalue = evalue
+        self.message = message
+        super().__init__(self.message)
+
+def evaluate_in_sage(query: str, format=False) -> str:
+    if format:
+        clean_query = query
+        query = (f'print({clean_query})')
+    data = {'code': query}
+    headers = {'content-type': 'application/x-www-form-urlencoded'}
+    response = requests.post('https://sagecell.sagemath.org/service', data, headers=headers)
+    response = json.loads(response.text)
+    if response['success']:
+        return response['stdout'] if 'stdout' in response else None
+    elif 'execute_reply' in response and 'ename' in response['execute_reply'] and 'evalue' in response['execute_reply']:
+        raise EvaluationError(response['execute_reply']['ename'], response['execute_reply']['evalue'])
+    else:
+        raise Exception(response)
+
+def main():
+    '''The system args contain the following:
+    0 - the path to this python file
+    1 - a string containing "tt" or "ff" depending on whether polyrith was called with trace enabled
+    2 - a string representing the base type of the target
+    3 - a list of all the variables to be used
+    4 - a list of the polynomial hypotheses/proof terms in terms of the variables
+    5 - a single polynomial representing the target'''
+    command = create_query(sys.argv[2], sys.argv[3], sys.argv[4], sys.argv[5])
+    final_query = polynomial_formatting_functions + "\n" + command
+    if sys.argv[1] == 'tt': # trace enabled
+        print(command)
+    else:
```

Comment:
Surely this makes everything fail if tracing is enabled?

Reply:
You're right, the tracing options need an overhaul. This is a bit tricky and there's time pressure, since I'd like to have this available at ICERM if possible. Maybe we can leave this for now and I'll write an IOU?

Reply:
I'm a little out of the loop; precisely when is that?

Reply:
It starts Monday (Jul 11). There will be a talk later on Tuesday when it would be convenient to have `polyrith` available!

Reply:
(Convenient, but not necessary, it's not a major deal.)